src/sentry/api/event_search.py

```diff
@@ -1365,233 +1365,354 @@ def get_default(self, params):
+        :param [str, [any], str or None] aggregate: The aggregate to be passed to snuba once formatted.
+            The arguments will be filled into the aggregate where needed. This must be an aggregate.
+        :param str transform: NOTE: Use aggregate over transform whenever possible.
+            An aggregate string to be passed to snuba once formatted. The arguments
+            will be filled into the string using `.format(...)`.
+        :param str result_type: The resulting type of this function. Can be any of the following
+            (duration, string, number, integer, percentage, date).
+        """
+
+        self.name = name
+        self.result_type = result_type
+        self.required_args = [] if required_args is None else required_args
+        self.optional_args = [] if optional_args is None else optional_args
+        self.calculated_args = [] if calculated_args is None else calculated_args
+        self.column = column
+        self.aggregate = aggregate
+        self.transform = transform
+
+        self.validate()
+
+    @property
+    def required_args_count(self):
+        return len(self.required_args)
+
+    @property
+    def optional_args_count(self):
+        return len(self.optional_args)
+
+    @property
+    def total_args_count(self):
+        return self.required_args_count + self.optional_args_count
+
+    @property
+    def args(self):
+        return self.required_args + self.optional_args
+
+    def format_as_arguments(self, field, columns, params):
+        # make sure to validate the argument count first to
+        # ensure the right number of arguments have been passed
+        self.validate_argument_count(field, columns)
+
+        columns = [column for column in columns]
+
+        # use default values to populate optional arguments if any
+        for argument in self.args[len(columns) :]:
+            try:
+                default = argument.get_default(params)
+            except InvalidFunctionArgument as e:
+                raise InvalidSearchQuery(u"{}: invalid arguments: {}".format(field, e))
+
+            # Hacky, but we expect column arguments to be strings so easiest to convert it back
+            columns.append(six.text_type(default) if default else default)
+
+        arguments = {}
+
+        # normalize the arguments before putting it in a dict
+        for column_value, argument in zip(columns, self.args):
+            try:
+                normalized_value = argument.normalize(column_value)
+                arguments[argument.name] = normalized_value
+            except InvalidFunctionArgument as e:
+                raise InvalidSearchQuery(
+                    u"{}: {} argument invalid: {}".format(field, argument.name, e)
+                )
+
+        # populate any computed args
+        for calculation in self.calculated_args:
+            arguments[calculation["name"]] = calculation["fn"](arguments)
+
+        return arguments
+
+    def validate(self):
```

Comment:
Since you are using a dict for all the arguments, you might want to check for duplicated names too. E.g. someone has two args with the same name.

Reply:
Ahh didn't want to change the implementation too much and missed that. I'll add that to the validation as well. 👍